src/FsAutoComplete.Core/Extensions.fs

```diff
@@ -24,6 +24,254 @@ type FSharpSymbol with
         | :? FSharpField as m -> not m.Accessibility.IsPublic
         | _ -> false
 
+type FSharpSymbolUse with
+    member this.IsPrivateToFile =
+        let isPrivate =
+            match this.Symbol with
+            | :? FSharpMemberOrFunctionOrValue as m -> not m.IsModuleValueOrMember || m.Accessibility.IsPrivate
+            | :? FSharpEntity as m -> m.Accessibility.IsPrivate
+            | :? FSharpGenericParameter -> true
+            | :? FSharpUnionCase as m -> m.Accessibility.IsPrivate
+            | :? FSharpField as m -> m.Accessibility.IsPrivate
+            | _ -> false
+
+        let declarationLocation =
+            match this.Symbol.SignatureLocation with
+            | Some x -> Some x
+            | _ ->
+                match this.Symbol.DeclarationLocation with
+                | Some x -> Some x
+                | _ -> this.Symbol.ImplementationLocation
+
+        let declaredInTheFile =
+            match declarationLocation with
+            | Some declRange -> declRange.FileName = this.RangeAlternate.FileName
+            | _ -> false
+
+        isPrivate && declaredInTheFile
+
+
+type FSharpMemberOrFunctionOrValue with
+
+    member x.IsConstructor = x.CompiledName = ".ctor"
+
+    member x.IsOperatorOrActivePattern =
+        let name = x.DisplayName
+        if name.StartsWith "( " && name.EndsWith " )" && name.Length > 4
+        then name.Substring (2, name.Length - 4) |> String.forall (fun c -> c <> ' ')
```

Comment:
the substring and length can be avoided. i really fear these.

you can use [chooseByPrefix](https://github.com/fsharp/FsAutoComplete/blob/6ee483f71935e93a988e30edf38929fe3ec60bb5/src/FsAutoComplete.Core/Utils.fs#L374-L376) to pattern match if start with that. and a add chooseBySuffix for the final